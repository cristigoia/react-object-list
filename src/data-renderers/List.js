import React, { Component } from 'react'
import PropTypes from 'prop-types'
import ClassNames from 'classnames'

import ListCard from './ListCard'
import { getVisibleColumns, setColumnLabels } from '../utils/functions'
import { DATA_TYPE } from '../utils/constants'

export default class ListRenderer extends Component {
  static propTypes = {
    columns: PropTypes.arrayOf(PropTypes.oneOfType([DATA_TYPE, PropTypes.arrayOf(DATA_TYPE)])),
    data: PropTypes.array,
    meta: PropTypes.object,
    // Custom component to be passed to render the list items
    Renderer: PropTypes.func,
    // Extra classes for the Renderer component to be rendered with
    extraClasses: PropTypes.string,
  }

  static defaultProps = {
    meta: {
      extraColumns: [],
      sortKeys: [],
    },
    columns: [],
    data: [],
    Renderer: ListCard,
  }

  constructor(props) {
    super(props)
    this.state = {
      columns: getVisibleColumns(setColumnLabels(this.props.columns), this.props.meta.extraColumns),
    }
  }

  componentWillReceiveProps(nextProps) {
    this.setState(() => ({columns: getVisibleColumns(setColumnLabels(nextProps.columns), nextProps.meta.extraColumns)}))
  }

  renderListRows = () => {
    const {Renderer, data} = this.props
    return data.map((row, idx) => {
      return (
        <Renderer
          key={`list-row-${idx}`}
          columns={this.state.columns}
          data={row}
        />
      )
    })
  }

  render() {
    return (
      <ul className={ClassNames('objectlist-list__list', this.props.extraClasses)}>
        {this.renderListRows()}
      </ul>
    )
  }
}
