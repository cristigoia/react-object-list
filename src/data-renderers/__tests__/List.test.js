import React from 'react'
import { shallow } from 'enzyme'
import List from '../List'

jest.mock('../../utils/functions', () => ({
  getVisibleColumns: (a, b) => [...a, ...b],
  setColumnLabels: input => input,
}))
jest.mock('../Overlay', () => 'Overlay')

describe('List', () => {
  describe('Lifecycle', () => {
    it('componentWillReceiveProps', () => {
      const newProps = {
        columns: [{dataKey: 'a'}, {dataKey: 'b'}],
        meta: {
          extraColumns: [{dataKey: 'c'}, {dataKey: 'd'}],
        },
      }
      const instance = shallow(<List />)
      instance.instance().setState({columns: []})
      instance.setProps(newProps)
      expect(instance.instance().state.columns).toEqual([...newProps.columns, ...newProps.meta.extraColumns])
    })
  })
})
