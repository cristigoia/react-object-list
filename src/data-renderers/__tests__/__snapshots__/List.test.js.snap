// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`List Snapshots can click on items 1`] = `
<div>
  <Overlay
    status="done"
  />
  <ul
    className="objectlist-list__list"
  >
    <li
      className="objectlist-list__item objectlist-list__item--clickable"
      onClick={[Function]}
    >
      <div
        className="objectlist-list__content"
      >
        <h2
          className="objectlist-list__header"
        >
          Name
          :
        </h2>
        <div
          className="objectlist-list__body"
        >
          
        </div>
      </div>
      <div
        className="objectlist-list__content"
      >
        <h2
          className="objectlist-list__header"
        >
          Age Gender
          :
        </h2>
        <div
          className="objectlist-list__body"
        >
           
        </div>
      </div>
    </li>
    <li
      className="objectlist-list__item objectlist-list__item--clickable"
      onClick={[Function]}
    >
      <div
        className="objectlist-list__content"
      >
        <h2
          className="objectlist-list__header"
        >
          Name
          :
        </h2>
        <div
          className="objectlist-list__body"
        >
          
        </div>
      </div>
      <div
        className="objectlist-list__content"
      >
        <h2
          className="objectlist-list__header"
        >
          Age Gender
          :
        </h2>
        <div
          className="objectlist-list__body"
        >
           
        </div>
      </div>
    </li>
    <li
      className="objectlist-list__item objectlist-list__item--clickable"
      onClick={[Function]}
    >
      <div
        className="objectlist-list__content"
      >
        <h2
          className="objectlist-list__header"
        >
          Name
          :
        </h2>
        <div
          className="objectlist-list__body"
        >
          
        </div>
      </div>
      <div
        className="objectlist-list__content"
      >
        <h2
          className="objectlist-list__header"
        >
          Age Gender
          :
        </h2>
        <div
          className="objectlist-list__body"
        >
           
        </div>
      </div>
    </li>
    <li
      className="objectlist-list__item objectlist-list__item--clickable"
      onClick={[Function]}
    >
      <div
        className="objectlist-list__content"
      >
        <h2
          className="objectlist-list__header"
        >
          Name
          :
        </h2>
        <div
          className="objectlist-list__body"
        >
          
        </div>
      </div>
      <div
        className="objectlist-list__content"
      >
        <h2
          className="objectlist-list__header"
        >
          Age Gender
          :
        </h2>
        <div
          className="objectlist-list__body"
        >
           
        </div>
      </div>
    </li>
    <li
      className="objectlist-list__item objectlist-list__item--clickable"
      onClick={[Function]}
    >
      <div
        className="objectlist-list__content"
      >
        <h2
          className="objectlist-list__header"
        >
          Name
          :
        </h2>
        <div
          className="objectlist-list__body"
        >
          
        </div>
      </div>
      <div
        className="objectlist-list__content"
      >
        <h2
          className="objectlist-list__header"
        >
          Age Gender
          :
        </h2>
        <div
          className="objectlist-list__body"
        >
           
        </div>
      </div>
    </li>
  </ul>
</div>
`;
